IntelFsp2Pkg/Library/BaseFspSwitchStackLib/Ia32/Stack.nasm

```diff
@@ -61,6 +85,58 @@ ASM_PFX(FspSwitchStack):
     ; Restore previous contexts
     lidt    [esp]
     add     esp, 8
+
+#if FixedPcdGetBool (EnableSaveRestorePageTable) == 1
+    ; Restore CR registers
+    pop     eax  ; new cr0
+    pop     ebx  ; new cr3
+    pop     ecx  ; new cr4
+
+    bt      eax, 31   ; if new cr0 is to enable page table,
+                      ; set the page table related control register first.
+    jc      SetPg
+    jmp     SkipSetPg
+
+SetPg:
+    mov     cr3, ebx
+
+    mov     edx, cr4
+    and     edx, ~CR4_PG_MASK
+    and     ecx, CR4_PG_MASK
+    or      edx, ecx
+    mov     cr4, edx
+
+SkipSetPg:
+    mov     edx, cr0
+    and     edx, ~CR0_PG_MASK
+    and     eax, CR0_PG_MASK
+    or      edx, eax
+    mov     cr0, edx
+
+    mov     cr3, ebx
+
+    mov     edx, cr4
+    and     edx, ~CR4_PG_MASK
+    and     ecx, CR4_PG_MASK
+    or      edx, ecx
+    mov     cr4, edx
+
+    ; Restore EFER MSR
+    pop    ebx
+    mov    ecx, 0xC0000080
+    rdmsr
+    and    eax, ~EFER_PG_MASK
+    and    ebx, EFER_PG_MASK
+    or     eax, ebx
+    wrmsr

```

Comment:
Setting NXE bit in EFER MSR in this late phase might be dangerous.

Reply:
Good catch. Will figure out a way to handle it.